warp10/src/main/java/io/warp10/script/functions/GUARD.java

```diff
@@ -114,8 +137,52 @@ public Object apply(WarpScriptStack stack) throws WarpScriptException {
       if (hasHide) {
         stack.show(hidden);
       }
+
+      // Restore capabilities
+
+      boolean copyExported = false;
+
+      if (exportedCapabilities instanceof AtomicReference) {
+        AtomicReference ref = (AtomicReference) exportedCapabilities;
+        if (ref.get() instanceof Set && ((Set) ref.get()).contains(null)) {
+          copyExported = true;
+        }
+      }
+
+      Capabilities caps = Capabilities.get(stack);
+      if (null != caps) {
+        // Export the specified capabilities if no error occurred
+        if (null == error) {
+          Capabilities newcaps = caps;
+          for (String cap: getExportedCapabilities(stack)) {
+            if (null != cap) {
+              if (copyExported) {
+                ((AtomicReference<Set>) exportedCapabilities).get().add(cap);
+              }
+              capabilities.remove(cap);
+              capabilities.putIfAbsent(cap, newcaps.get(cap));
+            }
+          }
+        }
+      }
+
+      if (null != capabilities) {
+        Capabilities.set(stack, capabilities);
```

Comment:
If there is no capability set with before the call to GUARD, all capabilities in macro will leak.